### DISCUSSION TITLE
Skupper expose - UDP Service

### DISCUSSION THREAD

**Discussion by vlanx** [skupperproject/skupper#1166] [Q&A] [answered]
Hi!
I was wondering what's the procedure if we want to expose a service/deployment that uses the UDP Protocol.
In the [documentation](https://skupper.io/docs/cli-reference/skupper_expose.html) i've only found option for HTTP/S and TCP.

Thanks in advance!

**Comment by jiridanek** [comment]
Skupper router does not support UDP. You can try to use the `--bridge-image` option and specify an udp bridge there. I never tried it, so I'd be myself interested in how it goes!

https://github.com/skupperproject/skupper/blob/d9aaf1c3849627f4a37064382d2f917520a431ae/cmd/skupper/skupper.go#L714

See example at https://github.com/skupperproject/skupper-test-udp-bridge for how it might be used.

>     skupper expose deployment udp-server --port 8000 --protocol udp --bridge-image=quay.io/skupper/bridge-server


The bridge itself lives at https://github.com/grs/skupper-udp-bridge.

**Reply by vlanx** [reply]
Hi @jiridanek, thank you for your help!
Although I followed the [example](https://github.com/skupperproject/skupper-test-udp-bridge) and it did work, with the logs of the udp server showing the messages from the udp client, when I followed the same steps for the my UDP service, sadly there was no connectivity. 

Services on VM A:
<pre>
~$ kubectl get svc -o wide
NAME                   TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                          AGE   SELECTOR
skupper                NodePort    192.168.87.157   <none>        8081:18941/TCP                   44m   app.kubernetes.io/name=skupper-service-controller, ...
skupper-router         NodePort    192.168.87.61    <none>        55671:6280/TCP,45671:14063/TCP   44m   app.kubernetes.io/name=skupper-router,app.kubernetes.io/part-of=skupper,...
skupper-router-local   ClusterIP   192.168.87.17    <none>        5671/TCP                         44m   app.kubernetes.io/name=skupper-router,app.kubernetes.io/part-of=skupper,...
<b>upf                    ClusterIP   192.168.87.236   <none>        8805/UDP                         41m   app=upf,release=upf1</b>
upf-http               ClusterIP   192.168.87.28    <none>        8000/TCP,8080/TCP                41m   app=upf,release=upf1
</pre>

I want to expose `upf` service, so I did the following:

```
skupper expose service upf --protocol udp --bridge-image=quay.io/skupper/bridge-server --address upf1
```

With the new service being created:
<pre>
~$ kubectl get svc -o wide

NAME                   TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                          AGE     SELECTOR
skupper                NodePort    192.168.87.157   <none>        8081:18941/TCP                   52m     app.kubernetes.io/name=skupper-service-controller,...
skupper-router         NodePort    192.168.87.61    <none>        55671:6280/TCP,45671:14063/TCP   52m     app.kubernetes.io/name=skupper-router,...
skupper-router-local   ClusterIP   192.168.87.17    <none>        5671/TCP                         52m     app.kubernetes.io/name=skupper-router,app.kubernetes.io/part-of=skupper,...
<b>upf                    ClusterIP   192.168.87.236   <none>        8805/UDP                         49m     app=upf,release=upf1</b>
upf-http               ClusterIP   192.168.87.28    <none>        8000/TCP,8080/TCP                49m     app=upf,release=upf1
<b>upf1                   ClusterIP   192.168.87.100   <none>        8805/UDP                         3m12s   skupper.io/external-bridge=upf1</b>
</pre>
After this `VM B` had the following services:
<pre>
~$ kubectl get svc

NAME                         TYPE        CLUSTER-IP       EXTERNAL-IP     PORT(S)                                       AGE
...
skupper                      NodePort    192.168.85.48    <none>          8010:9279/TCP,8080:16592/TCP,8081:20086/TCP   51m
skupper-prometheus           ClusterIP   192.168.85.71    <none>          9090/TCP                                      51m
skupper-router               NodePort    192.168.85.200   <none>          55671:21087/TCP,45671:20552/TCP               51m
skupper-router-local         ClusterIP   192.168.85.50    <none>          5671/TCP                                      51m
<b>smf                          ClusterIP   192.168.85.5     <none>          8805/UDP,9089/TCP,29502/TCP                   56m</b>
udm                          ClusterIP   192.168.85.142   <none>          29503/TCP                                     56m
udr                          ClusterIP   192.168.85.202   <none>          29504/TCP                                     56m
<b>upf1                         ClusterIP   192.168.85.31    <none>          8805/UDP                                      84s</b>
webui                        NodePort    192.168.85.168   <none>          5000:30002/TCP,9876:32581/TCP                 56m
</pre>
In my case, the `smf` is the one who will be connecting with the `upf1` service, and checking the logs I can see that the `smf` pod is sending packets to the correct service address, but nothing arrives in `VM A`.

On VM A, this is the logs output of the UDP Bridge, which are do not have the same output as the UDP Client-Server example:
```
~$ kubectl logs skupper-upf1-bridge-848589b8d8-2rg6k

create_udp_listener({ port: 8805, address: 'upf1_8805', multicast: undefined })
Healthz served on port 9090
```

I don't know how to debug further. Let me know if there's something more I should add!

**Reply by grs** [reply, marked as answer]
We would expect to see a connect being created on VM A. That is what will direct traffic at the intended target. What does your skupper-services configmap contain?

If you expose the deployment backing the service, instead of the service itself, does that make any difference?

**Reply by vlanx** [reply]
Hi @grs , thank you for your reply.
Making the change from exposing `service` to exposing `statefulset` solved the UDP bridge issue. Now the pods can communicate via UDP using Skupper. 

Thank you both for your help!
